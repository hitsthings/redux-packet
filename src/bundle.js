import { bindActionCreators } from "redux";

/*
//usage of bundle
bundler({
    fromContext: {
        selector: (state, contextA, contextB) => `itemsInContext`, // or a function returning this to make one per component
        actions: (contextA, contextB) => `actionsForContext`,
    }
})
const users = bundle({
    fromGroup: {
        selector: () => createStructuredSelector({
            users: usersForGroupSelector,
            usersLoading: usersLoadingForGroupSelector,
            usersAllLoaded: usersAllLoadedForGroupSelector,
            usersLoadingErrors: usersLoadingErrorsForGroupSelector,
        }),
        actions: (dispatch, group) => ({
            loadUsers: () => dispatch(loadUsersForGroup(group)),
            addUserToGroup: user => dispatch(addUserToGroup(user, group)),
        }), // OR a raw object like actions: { actionA: actionCreatorA, actionB: actionCreatorB }
    },
    byPermission: { ... }
})
*/

const memoizeFirstMapStateToProps = (selector, executeContextSelectors, firstState, firstProps, firstValue, firstContext) => {
    if (executeContextSelectors) {
        let inner = (state, props) => {
            inner = (state, props) => selector(state, ...executeContextSelectors(props));
            if (props !== firstProps) { // nothing salvageable
                return inner(state, props); 
            }
            if (state !== firstState) { 
                return selector(state, ...firstContext); // same props, can memoize contexts
            }
            return firstValue;// same props and state, can return first value
        };
        return (state, props) => inner(state, props);
    }
    let inner = state => {
        inner = selector;
        return state === firstState ?
            firstValue :
            inner(state);
    };
    return state => inner(state);
};

const checkBundleDescriptor = (bundleDescriptor, name, calledFnName, argExpectation) => {
    if (typeof bundleDescriptor !== 'object' || bundleDescriptor === null) {
        throw new TypeError(`${calledFnName} expects ${argExpectation}an object, but ${name} was not.`);
    }
    if (bundleDescriptor.selector && typeof bundleDescriptor.selector !== 'function') {
        throw new TypeError(`${calledFnName} expects selector to be a function, but ${name}.selector was not.`);
    }
    if (bundleDescriptor.actions && (typeof bundleDescriptor.actions !== 'function' && typeof bundleDescriptor.actions !== 'object')) {
        throw new TypeError(`${calledFnName} expects actions to be a function or object, but ${name}.actions was not.`);
    }
};

const checkBundleMapAndGetNames = bundleMap => {
    if (typeof bundleMap !== 'object' || bundleMap === null) {
        throw new TypeError('bundleAll(bundleMap) expects bundleMap to be an object.');
    }
    if (Array.isArray(bundleMap)) {
        throw new TypeError('bundleAll(bundleMap) expects bundleMap to be an object, not an Array.');
    }
    const bundleNames = Object.keys(bundleMap);
    if (bundleNames.length === 0) {
        throw new TypeError('bundleAll(bundleMap) expected to receive an object with properties, but found no properties.')
    }
    return bundleNames;
};

const selectorExecutor = selectors => {
    selectors = selectors.map(selector => typeof selector === 'function' ? selector : () => selector);
    return props => selectors.map(selector => selector(props));
};

const internalBundle = (bundleDescriptor, name) => {
    const Bundle = (...contextSelectors) => {
        const hasContext = contextSelectors.length !== 0;
        const executeContextSelectors = hasContext && selectorExecutor(contextSelectors);
        if ('minimumSelectorsExpected' in bundleDescriptor && contextSelectors.length < bundleDescriptor.minimumSelectorsExpected) {
            throw new Error(`The bundle ${name} expects at least ${bundleDescriptor.minimumSelectorsExpected} selectors to be provided.`);
        }
        const mapStateToProps = bundleDescriptor.selector ? (firstState, firstProps) => {
            const firstContext = hasContext ? executeContextSelectors(firstProps) : [];
            let firstValue = bundleDescriptor.selector(firstState, ...firstContext);
            if (typeof firstValue !== 'function') { // no factory, fully memoized first call
                return memoizeFirstMapStateToProps(bundleDescriptor.selector, executeContextSelectors, firstState, firstProps, firstValue, firstContext);
            }
            if (hasContext) { // factory, but memoize first context if we can. Passing in a new object as firstState means we'll never return a memoized firstValue
                return memoizeFirstMapStateToProps(firstValue, executeContextSelectors, {}, firstProps, null, firstContext);
            }
            // factory, no context, no memoization
            return state => firstValue(state);
        } : null;
        const mapDispatchToProps = typeof bundleDescriptor.actions === 'function' ?
            hasContext ?
                (dispatch, props) => bundleDescriptor.actions(dispatch, ...executeContextSelectors(props)) :
                dispatch => bundleDescriptor.actions(dispatch) :
            bundleDescriptor.actions ?
                dispatch => bindActionCreators(bundleDescriptor.actions, dispatch) :
                null;
        return {
            mapStateToProps,
            mapDispatchToProps,
        };
    };
    if (name) {
        Object.defineProperty(Bundle, 'name', {
            writable: false,
            enumerable: false,
            configurable: true,
            value: `Bundle(${name})`
        });
    }
    return Bundle;
};

export const bundle = (bundleDescriptor, name) => {
    checkBundleDescriptor(bundleDescriptor, "bundleDescriptor", "bundle(bundleDescriptor)", "");
    return internalBundle(bundleDescriptor, name);
};

export const bundleAll = bundleMap => {
    const bundleNames = checkBundleMapAndGetNames(bundleMap);
    return bundleNames.reduce((o,name) => {
        const bundleDescriptor = bundleMap[name];
        checkBundleDescriptor(bundleDescriptor, `bundleMap['${name}']`, "bundleAll(bundleMap)", "each property to be ");
        o[name] = internalBundle(bundleDescriptor, name);
        return o;
    }, {});
};