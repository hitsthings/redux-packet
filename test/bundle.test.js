import { bundleAll } from '../src/bundle.js';

describe('bundleAll()', () => {
    describe('contract assertions', () => {
        describe('throws on invalid input', () => {
            [
                undefined,
                null,
                true,
                false,
                0,
                1,
                '',
                'string',
                ['array'],
                {},
                {a: true},
                {a: 1},
                {a: 'string'},
                {a: {selector:true}},
                {a: {selector:1}},
                {a: {selector:'string'}},
                {a: {actions:true}},
                {a: {actions:1}},
                {a: {actions:'string'}},
            ].forEach(value =>
                it(String(JSON.stringify(value)), () => {
                    expect(() => bundleAll(value).toThrowErrorMatchingSnapshot())
                })
            );
        });
        describe('accepts valid input', () => {
            [
                { a:{} },
                { a:{}, b: {} },
                { a:{ selector: () => {}} },
                { a:{ actions: {} } },
                { a:{ actions: () => {} } },
            ].forEach(value => {
                it(String(JSON.stringify(value)), () => {
                    expect(() => bundleAll(value)).not.toThrow();
                })
            })
        });
        
        it('returns an object of functions with keys matching the input', () => {
            expect(bundleAll({ a: {}})).toEqual({ a: expect.any(Function) });
        });
    });

    describe('minimumSelectorsExpected', () => {
        it('does not validate if not passed in', () => {
            const bundled = bundleAll({ a: { }});
            expect(() => bundled.a()).not.toThrow();
        })
        it('validates contextSelector count', () => {
            const bundled = bundleAll({ a: { minimumSelectorsExpected: 1 }});
            expect(() => bundled.a()).toThrowErrorMatchingSnapshot();
            expect(() => bundled.a(() => {})).not.toThrow();
            expect(() => bundled.a(() => {}, () => {})).not.toThrow();
        });
    });

    describe('bundles', () => {
        describe('with context', () => {
            describe('simple', () => {
                let bundledWithContext, contextProviderA, makeSelector, selector, loadActionCreator;
                const contextValue = 40;
                const mappedProps = { prop: 42 };
                const loadAction = { type: 'A' };
                beforeEach(() => {
                    contextProviderA = jest.fn().mockReturnValue(contextValue);
                    selector = jest.fn().mockReturnValue(mappedProps);
                    loadActionCreator = jest.fn().mockReturnValue(loadAction);
                    bundledWithContext = bundleAll({
                        a: {
                            selector,
                            actions: {
                                load: loadActionCreator
                            },
                        }
                    }).a(contextProviderA);
                });
                it('returns all bundle props', () => {
                    expect(bundledWithContext).toEqual({
                        mapStateToProps: expect.any(Function),
                        mapDispatchToProps: expect.any(Function),
                    });
                });
                it('calls selector in makeMapStateToProps', () => {
                    const state = {}, props = {};
                    const mapStateToProps = bundledWithContext.mapStateToProps(state, props);
                    expect(selector).lastCalledWith(state, contextValue);

                    const mapped = mapStateToProps(state, props);
                    expect(mapped).toBe(mappedProps);

                    expect(selector).toHaveBeenCalledTimes(1);
                    expect(contextProviderA).toHaveBeenCalledTimes(1);
                });
                it('calls actions in mapDispatchToProps', () => {
                    const dispatch = jest.fn(), props = {};
                    const mapped = bundledWithContext.mapDispatchToProps(dispatch, props);
                    expect(mapped).toEqual({
                        load: expect.any(Function)
                    });
                    const loadParam = {};
                    const action = mapped.load(loadParam);
                    expect(loadActionCreator).lastCalledWith(loadParam);
                    expect(dispatch).lastCalledWith(loadAction);
                    expect(loadActionCreator).toHaveBeenCalledTimes(1);
                    expect(dispatch).toHaveBeenCalledTimes(1);
                    expect(contextProviderA).toHaveBeenCalledTimes(0);
                });
            });

            describe('complex', () => {
                let bundled, contextProviderA, makeSelector, selector, actions;
                const contextValue = 40;
                const mappedProps = { prop: 42 };
                const mappedActions = {
                    load: () => ({ type: 'A' })
                };
                beforeEach(() => {
                    contextProviderA = jest.fn().mockReturnValue(contextValue);
                    selector = jest.fn().mockReturnValue(mappedProps);
                    makeSelector = jest.fn().mockReturnValue(selector);
                    actions = jest.fn().mockReturnValue(mappedActions);
                    bundled = bundleAll({
                        a: {
                            selector: makeSelector,
                            actions,
                        }
                    });
                });
                it('returns all bundle props', () => {
                    expect(bundled.a(contextProviderA)).toEqual({
                        mapStateToProps: expect.any(Function),
                        mapDispatchToProps: expect.any(Function),
                    });
                });
                it('calls makeSelector and selector in makeMapStateToProps', () => {
                    const state = {}, props = {};
                    const mapStateToProps = bundled.a(contextProviderA).mapStateToProps(state, props);
                    expect(makeSelector).lastCalledWith(state, contextValue);
                    expect(selector).not.toHaveBeenCalled();

                    const mapped = mapStateToProps(state, props);
                    expect(selector).lastCalledWith(state, contextValue);
                    expect(mapped).toBe(mappedProps);

                    expect(makeSelector).toHaveBeenCalledTimes(1);
                    expect(selector).toHaveBeenCalledTimes(1);
                    expect(actions).not.toHaveBeenCalled();
                    expect(contextProviderA).toHaveBeenCalledTimes(1);
                });
                it('calls actions in mapDispatchToProps', () => {
                    const dispatch = jest.fn(), props = {};
                    const mapped = bundled.a(contextProviderA).mapDispatchToProps(dispatch, props);
                    expect(actions).lastCalledWith(dispatch, contextValue);
                    expect(mapped).toBe(mappedActions);

                    expect(actions).toHaveBeenCalledTimes(1);
                    expect(selector).not.toHaveBeenCalled();
                    expect(makeSelector).not.toHaveBeenCalled();
                    expect(contextProviderA).toHaveBeenCalledTimes(1);
                });
            });
        });

        describe('without context', () => {
            describe('simple', () => {
                let bundled, makeSelector, selector, loadActionCreator;
                const contextValue = 40;
                const mappedProps = { prop: 42 };
                const loadAction = { type: 'A' };
                beforeEach(() => {
                    selector = jest.fn().mockReturnValue(mappedProps);
                    loadActionCreator = jest.fn().mockReturnValue(loadAction);
                    bundled = bundleAll({
                        a: {
                            selector,
                            actions: {
                                load: loadActionCreator
                            },
                        }
                    }).a();
                });
                it('returns all bundle props', () => {
                    expect(bundled).toEqual({
                        mapStateToProps: expect.any(Function),
                        mapDispatchToProps: expect.any(Function),
                    });
                });
                it('calls selector in makeMapStateToProps', () => {
                    const state = {}, props = {};
                    const mapStateToProps = bundled.mapStateToProps(state, props);
                    expect(selector).lastCalledWith(state);

                    const mapped = mapStateToProps(state, props);
                    expect(mapped).toBe(mappedProps);

                    expect(selector).toHaveBeenCalledTimes(1);
                });
                it('calls actions in mapDispatchToProps', () => {
                    const dispatch = jest.fn(), props = {};
                    const mapped = bundled.mapDispatchToProps(dispatch, props);
                    expect(mapped).toEqual({
                        load: expect.any(Function)
                    });
                    const loadParam = {};
                    const action = mapped.load(loadParam);
                    expect(loadActionCreator).lastCalledWith(loadParam);
                    expect(dispatch).lastCalledWith(loadAction);
                    expect(loadActionCreator).toHaveBeenCalledTimes(1);
                    expect(dispatch).toHaveBeenCalledTimes(1);
                });
            });

            describe('complex', () => {
                let bundled, makeSelector, selector, actions;
                const mappedProps = { prop: 42 };
                const mappedActions = {
                    load: () => ({ type: 'A' })
                };
                beforeEach(() => {
                    selector = jest.fn().mockReturnValue(mappedProps);
                    makeSelector = jest.fn().mockReturnValue(selector);
                    actions = jest.fn().mockReturnValue(mappedActions);
                    bundled = bundleAll({
                        a: {
                            selector: makeSelector,
                            actions,
                        }
                    });
                });
                it('returns all bundle props', () => {
                    expect(bundled.a()).toEqual({
                        mapStateToProps: expect.any(Function),
                        mapDispatchToProps: expect.any(Function),
                    });
                });
                it('calls makeSelector and selector in makeMapStateToProps', () => {
                    const state = {}, props = {};
                    const mapStateToProps = bundled.a().mapStateToProps(state, props);
                    expect(makeSelector).lastCalledWith(state);
                    expect(selector).not.toHaveBeenCalled();

                    const mapped = mapStateToProps(state, props);
                    expect(selector).lastCalledWith(state);
                    expect(mapped).toBe(mappedProps);

                    expect(makeSelector).toHaveBeenCalledTimes(1);
                    expect(selector).toHaveBeenCalledTimes(1);
                    expect(actions).not.toHaveBeenCalled();
                });
                it('calls actions in mapDispatchToProps', () => {
                    const dispatch = jest.fn(), props = {};
                    const mapped = bundled.a().mapDispatchToProps(dispatch, props);
                    expect(actions).lastCalledWith(dispatch);
                    expect(mapped).toBe(mappedActions);

                    expect(actions).toHaveBeenCalledTimes(1);
                    expect(selector).not.toHaveBeenCalled();
                    expect(makeSelector).not.toHaveBeenCalled();
                });
            });
        });
    });
});