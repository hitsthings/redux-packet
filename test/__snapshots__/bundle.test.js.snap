// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundleAll() minimumSelectorsExpected validates contextSelector count 1`] = `"The bundle a expects at least 1 selectors to be provided."`;
